pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/OverrideBothEqualsAndHashcodeRule.java

```diff
@@ -42,6 +43,25 @@ public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
         return data;
     }
 
+    @Override
+    public Object visit(ASTAnonymousClassDeclaration node, Object data) {
+        if (node.isInterface()) {
+            return data;
+        }
+        super.visit(node, data);
+        if (!implementsComparable && containsEquals ^ containsHashCode) {
+            if (nodeFound == null) {
+                nodeFound = node;
+            }
+            addViolation(data, nodeFound);
+        }
+        implementsComparable = false;
+        containsEquals = false;
+        containsHashCode = false;
+        nodeFound = null;
```

Comment:
since the rule processes the code as it's written, and these are state vars shared for the whole processing, this will break on code such as:

```java
public class Foo {
  public boolean equals(Object o) { // containsEquals = true
    return false;
  }

 public void bar() {
    new Object{ // anonymous class declared
       public int hashCode() { return 42; } // containsHashCode = true
    } // anonymous class is ok (false negative), reset containsEquals and containsHashCode to false

  public int hashCode() { return 42; } // containsHashCode = true
 } // class has hashCode but no equals! (false positive)
```

The simplest way to fix this would be to remove the flags, remove the `Object visit(ASTMethodDeclaration node, Object data)` override, and directly navigate the AST here to match the equals / hashCode methods by doing:

```java
node.getOperations().filter(…)
```